### PR TITLE
Generate output using Doxygen / Breathe

### PR DESCRIPTION
**Pull request by pmaybank**
* Modify Doxygen config to pick up include directories recursively

* Add DeviceMem struct to API Reference guide

* Add classes that are used in Flash Attention kernel

### CODE REVIEW

**Comment by pmaybank** [comment]
Commands for setting up dependencies / requirements:

sudo apt-get install doxygen
pip install sphinx_rtd_theme
pip install breathe
pip install sphinxcontrib.bibtex
sudo apt install latexmk

File: docs/source/refs.bib [hunk 1]
```diff
@@ -0,0 +1,7 @@
+
+@article{dao2022flashattention,
+  title={Flashattention: Fast and memory-efficient exact attention with io-awareness},
+  author={Dao, Tri and Fu, Daniel Y and Ermon, Stefano and Rudra, Atri and R{\'e}, Christopher},
+  journal={arXiv preprint arXiv:2205.14135},
+  year={2022}
+}
```
Comment (geyyer) [inline docs/source/refs.bib:7]: Could you add a newline?

**Review by geyyer — DISMISSED** [review]
LGTM! Please add the newline, otherwise looks good!

File: docs/source/refs.bib (hunk 1 above)
Comment (pmaybank) [reply docs/source/refs.bib]: Done

**Review by geyyer — APPROVED** [review]
LGTM!

**Pull request merged**